shared/src/NewtonsoftJsonPayloadSerializer.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// A <see cref="Newtonsoft.Json.JsonConvert"/> <see cref="PayloadSerializer"/> implementation.
+    /// </summary>
+    public class NewtonsoftJsonPayloadSerializer : PayloadSerializer
+    {
+        /// <summary>
+        /// The Content Type string.
+        /// </summary>
+        internal const string ApplicationJson = "application/json";
+
+        /// <summary>
+        /// The default instance of this class.
+        /// </summary>
+        public static readonly NewtonsoftJsonPayloadSerializer Instance = new NewtonsoftJsonPayloadSerializer();
+
+        /// <inheritdoc/>
+        public override string ContentType => ApplicationJson;
+
+        /// <inheritdoc/>
+        public override string SerializeToString(object objectToSerialize)
+        {
+            return JsonConvert.SerializeObject(objectToSerialize);
+        }
+
+        /// <inheritdoc/>
+        public override T DeserializeToType<T>(string stringToDeserialize)
+        {
+            return JsonConvert.DeserializeObject<T>(stringToDeserialize);
+        }
+
+        /// <inheritdoc/>
+        public override T ConvertFromObject<T>(object objectToConvert)
+        {
+            if (objectToConvert == null)
+            {
+                return default;
+            }
+            return ((JObject)objectToConvert).ToObject<T>();
+        }
+
+        /// <inheritdoc/>
+        public override bool TryGetNestedObjectValue<T>(object objectToConvert, string propertyName, out T outValue)
```

Comment:
this method is to help with inspection of the json node for property operations. This node traversal will depend on the serializer defined, so it will need to be implemented by the serializer (eg: the node could be JsonObject (newtonsoft) or JsonElement for system.text).
For eg: when a device wants to fetch the value of a property under a component, 
```json
"thermostat1": {
	"__t": "c",
	"targetTemperature": 25
},
```

the sdk only knows that the value under "thermostat1" is an object. The type of the object depends on the serializer used to deserialize the response from the service (the twin received).

Reply:
Where does it take the component information? Or perhaps a sample would help understand how would this method be used.

We can sync on this offline if you think it's easier to explain.

Reply:
This information is something that the sdk would leverage internally when looking for a property nested under a component:
eg: https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/pnpImplDev/iothub/device/src/ClientPropertyCollection.cs#L235
eg: https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/pnpImplDev/iothub/device/src/ClientPropertyCollection.cs#L262

Ping me if this is still unclear and we can discuss it offline.